pages/api/search.ts

```diff
@@ -0,0 +1,34 @@
+import axios from 'axios'
+import type { NextApiRequest, NextApiResponse } from 'next'
+
+import { encodePath, getAccessToken } from '.'
+import apiConfig from '../../config/api.json'
+
+export default async function handler(req: NextApiRequest, res: NextApiResponse) {
+  // Get access token from storage
+  const accessToken = await getAccessToken()
+
+  // Query parameter from request
+  const { q: searchQuery = '' } = req.query
+
+  if (typeof searchQuery === 'string') {
+    // Construct Microsoft Graph Search API URL, and perform search only under the base dir
+    const encodedPath = encodePath('/') === '' ? encodePath('/') : encodePath('/') + ':'
+    const searchApi = `${apiConfig.driveApi}/root${encodedPath}/search(q='${encodeURIComponent(searchQuery)}')`
+
+    try {
+      const { data } = await axios.get(searchApi, {
+        headers: { Authorization: `Bearer ${accessToken}` },
+        params: {
+          select: 'id,name,file,folder,parentReference',
+        },
+      })
+      res.status(200).json(data.value)
```

Comment:
Since there are protected routes, the returned pathes should be checked whether they are protected with any `.password` if protected routes are still going to work on searching.
If we just exclude protected routes from searching, things are simple. Otherwise, the way to provide multiply passwords (as results may be protected by different `.password`, while currently only one password is stored and provided) seems difficult and requires consideration.

Reply:
Ok. I see what you mean, but I think we have to ignore this for now.

So basically, there are three places where we can filter out protected routes:

1. The direct search result returned from OneDrive's API - fetched on the server side, and **not** controlled by us
2. The search result returned to the client by our serverless API - returned also on the server side but controlled by us
3. The result rendered into a React component - using data fetched from our serverless API, also controlled by us

The only real 'filter process' should be performed on the server side, i.e., before the data gets to the client, meaning that we need to filtered out protected routes in the results in either 1 or 2.

Additionally, the routes that are 'decrypted' are stored in localStorage and not synced across multiple instances.

- We can rule out 1 as we cannot simply pass an argument to the OneDrive API and expect it to respect our defined protected routes.
- We can also rule out 3 as filtering on the client side is simply not doing any filtering at all.

So that leaves us with 2, which may seem promising, but in reality we still need our server side API to know which routes are 'decrypted'. This information is only stored in localStorage so we can either:

- 🚫 Add another query param specifying which routes are decrypted when requesting a search - but that won't work as one can simply query our search endpoint manually, passing all routes as 'decrypted', and getting a full list of results with protected routes
- 🚫 Sync the decrypted routes in Redis by session or user - which basically means we need to handle user login sessions and permissions in our project separately, and **no, I don't want to** 🙃 

So, I think I'm going to ignore this for now. (And while we get protected routes in the search results, the user would still have to enter the password to get access to the file no matter what - this only exposes the file names of the protected folders, which I think is acceptable?)